### PR TITLE
feat: t12 - 500 error handling

### PR DESCRIPTION
**Pull request by rtest12**
## t12 - 500 error handling
In the example I'm using when there is an error happening in the python code, you will get it logged (along with the traceback details) and the following error (handled by the python / uwsgi) will be displayed:
error1

So it didn't get handled by the nginx, what we want to do is to modify the solution in such a way so that we get any 50x errors handled by nginx instead, so when you hit it, you get error handled by nginx:

error2

use `50x.html` which is in the `l3` folder to handle such 50x errors.

When you open PR - pls package your solution as zip file with bash script `run.sh` that should

-    create all necessary files and/or folders.

-    set proper permissions to created files/folders

-    start nginx in foreground mode (so all errors, logs are printed on stdout/stderr)

-    start uwsgi in foreground mode (so all errors, logs are printed on stdout/stderr)

-    http GET https://localhost/changelog should result in the proper HTML code as on the picture

-    http POST https://localhost/changelog changelog="lecture about pycamp" should succeed with the following response:
```
    HTTP/1.1 200 OK
    Connection: keep-alive
    Content-Length: 44
    Content-Type: application/json
    Date: Thu, 13 Apr 2023 00:29:38 GMT
    Server: nginx/1.22.1
    x-server-engine: pytnon

    "Successfully stored: lecture about pycamp"
```

## My zip

[Task_12.zip](https://github.com/rtest12/devops-camp-2023/files/11346909/Task_12.zip)

### CODE REVIEW

File: chapter2/tests/t12/nginx.conf [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
+					  '$status $body_bytes_sent "$http_referer" '
+					  '"$http_user_agent" "$http_x_forwarded_for"';
+
+	access_log  /tmp/camp/access.log main;
+	keepalive_timeout   65s;
+	types_hash_max_size 4096;
+	include             /etc/nginx/mime.types;
+	default_type        application/octet-stream;
+		index index.html;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t12/nginx.conf]: format file properly. if you keep columns starting in the same position - then do it.
use tabs vs spaces.

File: chapter2/tests/t12/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+events {
+	worker_connections 1024;
+}
+
+http {
+
+	log_format  main  '$remote_addr - $remote_user [$time_local] "$request" '
+					  '$status $body_bytes_sent "$http_referer" '
+					  '"$http_user_agent" "$http_x_forwarded_for"';
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t12/nginx.conf]: find a way to format this better.

File: chapter2/tests/t12/nginx.conf [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+	access_log  /tmp/camp/access.log main;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t12/nginx.conf]: group directives by purpose. Your error_log is at the very top. Same should be for the access_log

File: chapter2/tests/t12/nginx.conf [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		server_name  _;
+		root         /tmp/nginx;
+		ssl_certificate "/tmp/nginx/certs/cert.pem";
+		ssl_certificate_key "/tmp/nginx/certs/key.pem";
+		ssl_session_cache shared:SSL:1m;
+		ssl_session_timeout  10m;
+		ssl_ciphers PROFILE=SYSTEM;
+		ssl_protocols TLSv1.2 TLSv1.3;
+        ssl_ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384;
+        ssl_prefer_server_ciphers on;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t12/nginx.conf]: format issue.

File: chapter2/tests/t12/nginx.conf [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+		error_page 500 502 503 504 /50x.html;
+
+		location = /50x.html {
+		}		
+
+		location / {
+			add_header x-server-engine      "pytnon";
+			# Handling uwsgi errors with nginx
+			# http://nginx.org/en/docs/http/ngx_http_uwsgi_module.html#uwsgi_intercept_errors
+			uwsgi_intercept_errors on;
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t12/nginx.conf:64]: that is correct solution.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
solution is correct, but formatting is ugly. you can do better Max.
learn a habit - check your code in the PR prior assigning it to reviewers. That will help.

**Comment by dmitry-mightydevops** [comment]
your formatting of the nginx is
image

**Comment by rtest12** [comment]
> solution is correct, but formatting is ugly. you can do better Max. learn a habit - check your code in the PR prior assigning it to reviewers. That will help.

@dmitry-mightydevops 
Okay, I'll do my best, done

File: chapter2/tests/t12/nginx.conf (hunk 1 above)
Comment (rtest12) [reply chapter2/tests/t12/nginx.conf]: > format file properly. if you keep columns starting in the same position - then do it. use tabs vs spaces.


@dmitry-mightydevops 
Done

File: chapter2/tests/t12/nginx.conf (hunk 2 above)
Comment (rtest12) [reply chapter2/tests/t12/nginx.conf]: > find a way to format this better.

@dmitry-mightydevops 
Done

File: chapter2/tests/t12/nginx.conf (hunk 3 above)
Comment (rtest12) [reply chapter2/tests/t12/nginx.conf]: > group directives by purpose. Your error_log is at the very top. Same should be for the access_log

@dmitry-mightydevops 
Done

File: chapter2/tests/t12/nginx.conf (hunk 4 above)
Comment (rtest12) [reply chapter2/tests/t12/nginx.conf]: > format issue.

@dmitry-mightydevops 
Done

File: chapter2/tests/t12/nginx.conf (hunk 2 above)
Comment (rtest12) [reply chapter2/tests/t12/nginx.conf]: > find a way to format this better.

@dmitry-mightydevops 

I partially corrected it, but not the way I wanted to. The display behaves very strangely in Github, while in VSCode everything looks normal. I haven't figured out the cause yet, but from now on I'll double-check with my eyes after every push..
Screenshot from 2023-04-27 23-25-46

File: chapter2/tests/t12/nginx.conf [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+events {
+	worker_connections 1024;
+}
+
+http {
+
+	log_format main '$remote_addr - $remote_user [$time_local] "$request" '
+	'$status $body_bytes_sent "$http_referer" '
+	'"$http_user_agent" "$http_x_forwarded_for"';
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t12/nginx.conf]: that is not any better :)

```
log_format json_combined escape=json
  '{'
    '"time_local":"$time_local",'
    '"remote_addr":"$remote_addr",'
    '"remote_user":"$remote_user",'
    '"request":"$request",'
    '"status": "$status",'
    '"body_bytes_sent":"$body_bytes_sent",'
    '"request_time":"$request_time",'
    '"http_referrer":"$http_referer",'
    '"http_user_agent":"$http_user_agent"'
  '}';
```

File: chapter2/tests/t12/nginx.conf [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+		ssl_session_timeout 10m;
+		ssl_ciphers PROFILE=SYSTEM;
+		ssl_protocols TLSv1.2 TLSv1.3;
+		ssl_ciphers ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:\
+		ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:\
+		ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384;
+		ssl_prefer_server_ciphers on;
+		error_page 404 /404.html;
+		location = /404.html {
+		}
```
Comment (dmitry-mightydevops) [inline chapter2/tests/t12/nginx.conf:55]: prefer to have space before any location/server/upstream contexts - as they're important

**Review by dmitry-mightydevops — APPROVED** [review]

File: chapter2/tests/t12/nginx.conf (hunk 6 above)
Comment (rtest12) [reply chapter2/tests/t12/nginx.conf]: > 

@dmitry-mightydevops 
Thanks, done

File: chapter2/tests/t12/nginx.conf (hunk 7 above)
Comment (rtest12) [reply chapter2/tests/t12/nginx.conf:55]: > prefer to have space before any location/server/upstream contexts - as they're important

@dmitry-mightydevops 
Done

**Pull request merged**